link/struct_ops.go

```diff
@@ -0,0 +1,46 @@
+package link
+
+import (
+	"fmt"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal/sys"
+)
+
+type structOpsLink struct {
+	*RawLink
+}
+
+// AttachStructOps attaches a struct_ops map (created from a ".struct_ops.link"
+// section) to its kernel subsystem via a BPF link.
+func AttachStructOps(m *ebpf.Map) (Link, error) {
+	if m == nil {
+		return nil, fmt.Errorf("map cannot be nil")
+	}
+
+	if t := m.Type(); t != ebpf.StructOpsMap {
+		return nil, fmt.Errorf("invalid map type %s, expected struct_ops", t)
+	}
+
+	mapFD := m.FD()
+	if mapFD <= 0 {
+		return nil, fmt.Errorf("invalid map: %s (was it created?)", sys.ErrClosedFd)
+	}
+
+	// ".struct_ops.link" requires the map to be created with BPF_F_LINK.
+	if (int(m.Flags()) & sys.BPF_F_LINK) != sys.BPF_F_LINK {
```

Comment:
I'm guessing that the kernel does this check already? I'd probably drop this, since it doesn't matter from the library side whether the flag is set (it does matter that we don't create ones without, but that is already taken care of.)

Reply:
fixed